Standalone Egress NAT

p-strusiewiczsurmacki-mobica opened the pull request:
This PR introduces Standalone Egress NAT as discussed in #274 .

- `coil-controller` is now divided into `coil-ipam-controller` and `coil-egress-controller`.
- `coild` has now configuration flags to disable/enable egress and/or IPAM features.
~~- `coil` has now `capabilities` fields that can be used to disable/enable IPAM/Egress.~~
- If IPAM is disabled, coild can be run as secondary CNI using CNI chaining as described in `setup.md`.
- In case IPAM is disabled, `veth` aliases will now use pod's UUID instead of container's ID (couldn't get e2e test working using container's ID, I believe container ID is changed during pod restart, but as IPAM is disabled it is not updated as required).

PR was tested using egress related E2E tests with both Kindnet and Calico and tests that are provided in the repository passed.

terassyi commented:
@p-strusiewiczsurmacki-mobica 
I'm reviewing this, so please wait a little longer.

terassyi commented:
> PR was tested using egress related E2E tests with both Kindnet and Calico and tests that are provided in the repository passed.

Thank you for checking it.
I also ran stand-alone egress mode with Kindnet, and it works well.

It's better if we can run e2e tests for stand-alone egress mode.
What do you think?

terassyi commented:
Hi, @p-strusiewiczsurmacki-mobica !

Thank you for waiting.

The main changes look good to me.
So I left some nit comments.
Please check it.

p-strusiewiczsurmacki-mobica commented on v2/config/pod/coil-egress-controller.yaml:
```diff
@@ -0,0 +1,90 @@
+      labels:
+        app.kubernetes.io/component: coil-egress-controller
+    spec:
+      hostNetwork: true
```
Done

p-strusiewiczsurmacki-mobica commented on v2/controllers/egress_watcher.go:
```diff
@@ -24,6 +24,7 @@ type EgressWatcher struct {
 	NodeName   string
 	PodNet     nodenet.PodNetwork
 	EgressPort int
+	Logger     logr.Logger
```
Done

p-strusiewiczsurmacki-mobica commented on v2/pkg/nodenet/pod.go:
```diff
@@ -422,12 +454,21 @@ func (pn *podNetwork) Update(podIPv4, podIPv6 net.IP, hook SetupHook) error {
+				}
+			}
+		} else {
+			if c.ContainerId == string(pod.UID) {
```
I've added comment:

```
// We need to bind PodNetConf with Pod. When Coil is a primary CNI IP address can be used for that.
// However different CNIs manage IP addresses differently, therefore, in egress-only mode instead of
// storing container's ID in ContainerId, we store pod's UID there to be able to identify pod that was
// used to create particular PodNetConf.
```

Please let me know if this is what you were thinking about.

terassyi replied:
Thank you.
This is what I want!

p-strusiewiczsurmacki-mobica commented:
> > PR was tested using egress related E2E tests with both Kindnet and Calico and tests that are provided in the repository passed.
> 
> Thank you for checking it. I also ran stand-alone egress mode with Kindnet, and it works well.
> 
> It's better if we can run e2e tests for stand-alone egress mode. What do you think?

I agree this would be great. However I need to think about how this test should be done. The problem with kindnet is, that it requires different CNI config on each node, so I would need to add some script that would get node subnet from the kind cluster's nodes after the cluster was created and use that data to update CNI config after `make coil-install` is executed.

terassyi commented:
> I would need to add some script that would get node subnet from the kind cluster's nodes after the cluster was created and use that data to update CNI config after make coil-install is executed.

Yes, it isn't easy to configure the cluster to use kindnet and coil.

OK, in this PR, it's OK no e2e test for egress-only mode.
Thanks.

p-strusiewiczsurmacki-mobica commented on v2/runners/coild_server.go:
```diff
@@ -189,83 +192,172 @@ func (s *coildServer) Add(ctx context.Context, args *cnirpc.CNIArgs) (*cnirpc.Ad
+			return nil, newInternalError(err, "check failed")
+		}
+	}
+
```
I've added basic check for interface alias (as it was there before my changes and it was miising now in egress-only mode).
I was thinking about adding further checks, but it seems that to use e.g. IsInitalized functions of `founat` package we would need to get pod, service and egress from KubeAPI each time check is called, so it would need to be considered if this is worth it.
I think, that additional checks could be added, but I would rather have those as separate PR as this one is already kinda big.

terassyi replied:
Thanks.
I agree that we consider this later.
I think coil's CNI check command needs more consideration(not only egress-only mode, also ipam mode).
Therefore, this implementation is enough for now.

p-strusiewiczsurmacki-mobica commented:
@terassyi I've added egress-only e2e tests to `e2e/Makefile` and instructions on how to use those to `docs/setup.md`.

terassyi commented on v2/e2e/coil_test.go:
How about dividing egress related tests and ipam related ones into functions and using Context() to make them easy to understand?

p-strusiewiczsurmacki-mobica replied:
Done :)

terassyi commented on docs/setup.md:
```diff
@@ -231,3 +236,84 @@ host directory.
+### Testing standalone egress
+
+1. Configure Coil to run in standalone egress mode as described in [configuration](#configuration)
+1. Set proper CNI netconf to use in `e2e/kustomization.yaml`.
```
In the future, we want to run this test in CI.
Therefore, we want to reduce manual procedures.

How about preparing `kustomization.yaml` for each CNI setting and switching the environment to use with an env value?

p-strusiewiczsurmacki-mobica replied:
I've created separate kustomization YAMLs and added proper targets to the test Makefile. Instructions in `docs/setup.md` were updated accordingly.

p-strusiewiczsurmacki-mobica commented on v2/e2e/Makefile:
```diff
@@ -29,14 +37,44 @@ install-coil:
+	$(KUBECTL) label nodes coil-worker test=coil
+	$(KUBECTL) label nodes coil-worker2 test=coil
+	$(KIND) load docker-image --name coil coil:dev
+	$(KUSTOMIZE) build --load-restrictor=LoadRestrictionsNone . | $(KUBECTL) apply -f -
```
I've automated the process so this should not be an issue anymore.

terassyi commented on v2/e2e/coil_test.go:
For more advanced tests, we have to check if the egress NAT pod correctly proxies egress traffic.

Because when using kindnet, all pods belong to the same network segment and can communicate with external networks.
In this case, only checking the connectivity with external networks, we cannot strictly test the egress NAT feature. 

To test this, we have to check in the server(which is in the external network, like echo-server), the source address is NAT pod's one.

To get this type of test, we have to update echo-server to get the source address.

p-strusiewiczsurmacki-mobica replied:
I've added `req.RemoteAddr` to the `echo-server` reponse if `--reply-remote` flag is set. This will be used only in standalone-egress tests for  now.

The response is then compared against the list of NAT service endpoints.

p-strusiewiczsurmacki-mobica commented on v2/e2e/Makefile:
```diff
@@ -30,34 +32,76 @@ stop: $(KIND)
+	./kindnet-conf set 10-coil.conflist v6
 	rm -rf tmp kindnet-conf
 
+.PHONY: install-coil-egress-calico-v4
```
OK, I've deleted all Calico references.

p-strusiewiczsurmacki-mobica commented:
@terassyi Just one important thing - it turned out that `egress-controller` pods has to use `hostNetwork` as, otherwise, it seems to be unable to reach the apiserver if Coild is used as primary CNI.
So I've added the `hostNetwork` to the manifest once again.

terassyi commented:
@p-strusiewiczsurmacki-mobica 
Thank you for reporting.

> So I've added the hostNetwork to the manifest once again.

In this PR, it's ok to turn on `hostNetwork`.


When I tried to turn out `hostNetwork`, I found that `egess-controller` couldn't communicate with the api-server.
I'm not sure why it can't reach api-server.

But, I think `hostNetwork` is an unnecessary setting.
So, I will create an investigation issue after merging this.

terassyi commented:
https://github.com/cybozu-go/coil/pull/299/files#diff-87c0c48de880a4de757f70cabed8e310b213adaebd0974c60386af47de4aeb8bR81

Could you update the `manifests` target to be able to select generating manifests for only ipam or egress related or both?

p-strusiewiczsurmacki-mobica commented:
> https://github.com/cybozu-go/coil/pull/299/files#diff-87c0c48de880a4de757f70cabed8e310b213adaebd0974c60386af47de4aeb8bR81
> 
> Could you update the `manifests` target to be able to select generating manifests for only ipam or egress related or both?

@terassyi Done. :)

 `manifests`, `manifests-egress` and `manifests-ipam` are now available to be used.

It seems that controller-gen only accepts directories for manifest generation, so I had to add some workaround for that (copying `*_webhook.go` files to temporary folder) but I think it should work OK.

p-strusiewiczsurmacki-mobica commented:
@terassyi 
> ```
> Error: accumulating resources: accumulation err='accumulating resources from '../../../../config/default/egress/v4': read /home/terashima/go/src/github.com/cybozu-go/coil/v2/config/default/egress/v4: is a directory': recursed accumulation of path '/home/terashima/go/src/github.com/cybozu-go/coil/v2/config/default/egress/v4': trouble configuring builtin PatchStrategicMergeTransformer with config: `
> paths:
> - ../../webhook_manifests_patch.yaml
> `: open /home/terashima/go/src/github.com/cybozu-go/coil/v2/config/default/webhook_manifests_patch.yaml: no such file or directory
> error: no objects passed to apply
> ```

I believe you're missing a file `v2/config/default/webhook_manifests_patch.yaml` which is generated by `make certs` target (this is something i did not change).

So, you'd have to run 

```
make -C .. manifests && make -C .. certs
WITH_KINDNET=true TEST_IPV6=false make start
make install-coil-egress-v4
```

terassyi commented:
@p-strusiewiczsurmacki-mobica 

Thanks!
I could run e2e tests for each configuration, and it seems all e2e tests work well.

It seems procedures to run e2e tests are getting complex, so cloud you update `e2e/README.md`?

terassyi commented:
It seems webhook-related tests in small test fail.
Could you fix this test?

p-strusiewiczsurmacki-mobica commented:
@terassyi Small test should be fixed and actions were added to the CI.

Should I squash the commits before it'll be merged?

terassyi commented:
Thanks!

It passes all CIs.

> Should I squash the commits before it'll be merged?

Yes, please.

p-strusiewiczsurmacki-mobica commented:
@terassyi It's squashed now :)

terassyi commented:
@p-strusiewiczsurmacki-mobica 

Thanks!

I found small test seems to be flaky.
Cloud you fix this?

https://github.com/cybozu-go/coil/actions/runs/11663759016

p-strusiewiczsurmacki-mobica commented:
@terassyi I've fixed the warnings and added some fixes for the CI jobs.
One last thing that is bugging me are the controller-runtime's/go-client's cache errors in the `small tests`, like this one:

```
W1105 08:58:09.635773   24935 reflector.go:470] pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/reflector.go:232: watch of *v1.PodDisruptionBudget ended with: an error on the server ("unable to decode an event from the watch stream: context canceled") has prevented the request from succeeding
```

But I believe that this might be out of our reach here, as it seems it is [somewhat known issue with controller runtime's cache](https://github.com/kubernetes-sigs/controller-runtime/issues/2723).

I've tried disabling the cache altogether but for some reason it did not work for me.

If you see anything else I could fix just let me know.

RefluxMeds commented:
Hi @terassyi,
Have you had time to check this? It would be very beneficial for our org.
Thanks!

terassyi commented:
@RefluxMeds 

Hi, I'm checking this to be able to merge.
Please wait a little longer!

terassyi commented:
@p-strusiewiczsurmacki-mobica

After re-reviewing all changes, I want to discuss the need for `enable-spam/egress` flags in the coil CNI plugin.

I think it's enough to handle these flags in coild.

Now, we check the features we  want to use by values given by the coil CNI plugin.
However, could also has the same flags.
I think we can use coild's `enable-ipam/egress` flags instead of coil CNI's flags.

https://github.com/cybozu-go/coil/pull/299/files#diff-f7d3c7316366cfd19ef46b412e3ab6937dc7d6607b49a73b613e3fecd639d875R181

The only concern I have now is following the code.

https://github.com/cybozu-go/coil/pull/299/files#diff-e193be60a714369d09504366326bc0f86dff95f1dabd9e1bd16e027dac81acb2R33

But we can solve this by moving this error handling to coild.

If we can keep the CNI configuration simple and no changes, it's better for all users.

p-strusiewiczsurmacki-mobica commented:
@terassyi 
OK, I've removed the capabilities from the CNI binary itself. The configuration will be now handled by `coild` flags. I've also moved the check you've mentioned to `coild` and changed `small tests` a bit to reflect that.

terassyi commented on v2/runners/coild_server.go:
```diff
@@ -170,105 +179,210 @@ func newInternalError(err error, msg string) error {
-			"missing pod name/namespace", fmt.Sprintf("%+v", args.Args))
+	isChained, err := getSettings(args)
+	if err != nil {
+		return nil, newInternalError(fmt.Errorf("runtime error"), "failed to get CNi arguments")
```
```suggestion
		return nil, newInternalError(fmt.Errorf("runtime error"), "failed to get CNI arguments")
```

terassyi commented on v2/runners/coild_server.go:
```diff
@@ -88,7 +92,8 @@ func (n natSetup) Hook(l []GWNets, log *zap.Logger) func(ipv4, ipv6 net.IP) erro
 // NewCoildServer returns an implementation of cnirpc.CNIServer for coild.
-func NewCoildServer(l net.Listener, mgr manager.Manager, nodeIPAM ipam.NodeIPAM, podNet nodenet.PodNetwork, setup NATSetup, logger *zap.Logger) manager.Runnable {
+func NewCoildServer(l net.Listener, mgr manager.Manager, nodeIPAM ipam.NodeIPAM, podNet nodenet.PodNetwork, setup NATSetup, cfg *config.Config, logger *zap.Logger,
+	aliasFunc func(interfaces map[string]bool, conf *nodenet.PodNetConf, logger *zap.Logger, pod *corev1.Pod) error) manager.Runnable {
```
Why did you add the callback function to add the interface alias?

p-strusiewiczsurmacki-mobica replied:
So, in egress only mode we need to set interface alias on `Add` call. However the [function](https://github.com/p-strusiewiczsurmacki-mobica/coil/blob/standalone-egress/v2/runners/coild_server.go#L279) that sets alias is dependent on the existance of the pod's veth pair which is not created for `simpe-test`. I've added the callback function so `nil` returning function can be passed [here](https://github.com/p-strusiewiczsurmacki-mobica/coil/blob/standalone-egress/v2/runners/coild_server_test.go#L225) which makes it possible to run `simple-test` for egress-only mode.

At least that's the best I was able to come up with.

terassyi replied:
Thanks.
It's ok to me!

terassyi commented on v2/runners/coild_server.go:
```diff
@@ -170,105 +179,210 @@ func newInternalError(err error, msg string) error {
-	if err != nil {
-		logger.Sugar().Errorw("failed to allocate address", "error", err)
-		return nil, newInternalError(err, "failed to allocate address")
+	if !s.cfg.EnableIPAM && !s.cfg.EnableEgress {
```
I think it is enough to check only when starting the cold server.

p-strusiewiczsurmacki-mobica replied:
You're right! Done :) Moved that check to `v2/cmd/coild/run.go` (`subMain` function).

terassyi commented on v2/runners/coild_server.go:
```diff
@@ -170,105 +179,210 @@ func newInternalError(err error, msg string) error {
+	if err != nil {
+		return fmt.Errorf("netlink: failed to look up the host-side veth [%s]: %w", ifName, err)
+	}
+	logger.Sugar().Infof("link found: %v", hLink)
```
We can remove this log.

p-strusiewiczsurmacki-mobica replied:
Done :)

terassyi commented:
@p-strusiewiczsurmacki-mobica 
Thank you for fixing!

I added some comments :)
Please check it!

p-strusiewiczsurmacki-mobica commented:
Hi @terassyi - sorry you had to wait so long. I was on a leave for some time.

Anyway - I've fixed the issues you've found and added answer for your question regarding the callback function. I'll tests the changes tomorrow, but can I ask you to trigger the test workflow anyway?

terassyi commented:
Hi, @p-strusiewiczsurmacki-mobica 
thank you for updating!

I approved to run CI👍

p-strusiewiczsurmacki-mobica commented:
That's weird. I see that the small tests failed in the CI, but I can't see that error to happen on my machine.

terassyi reviewed (commented):
Thank you for updating!

I reviewed the whole changes and added some small comments.
After reflecting on these, we can merge this!

terassyi commented on .github/workflows/ci.yaml:
```diff
@@ -72,3 +81,56 @@ jobs:
+    name: End-to-end Test - Egress Only
+    strategy:
+      matrix:
+        kindest-node: ["1.27.13", "1.28.9", "1.29.4"]
```
Since latest version, coil supports kubernetes 1.30.
Cloud you update supported kubernetes versions for tests?

p-strusiewiczsurmacki-mobica replied:
Done :)

p-strusiewiczsurmacki-mobica commented on v2/Makefile:
```diff
@@ -78,36 +79,64 @@ check-generate:
 .PHONY: manifests
-manifests: $(CONTROLLER_GEN) $(ROLES) $(YQ)
-	$(CONTROLLER_GEN) $(CRD_OPTIONS) webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+# manifests: $(CONTROLLER_GEN) $(ROLES) $(YQ)
```
Done :)

p-strusiewiczsurmacki-mobica commented on .github/workflows/ci.yaml:
```diff
@@ -72,3 +81,56 @@ jobs:
+    - run: make install-coil-egress-v6
+      if: matrix.ip-version == 'ipv6'
+      working-directory: v2/e2e
+    - run: docker exec -i coil-worker ls /etc/cni/net.d/
```
Done :)

In fact, I've reworked that a bit and combined separate e2e-egress tests with 'full' tests using some conditional statements. I think that will be easier to handle if CI changes will be required in the future, instead of handling 2 separate instances of almost identical code.

terassyi replied:
Thanks!

terassyi commented:
@p-strusiewiczsurmacki-mobica 
Thanks a lot!
I approved this.


For merging, I'm testing compatibility.
Because, from this PR, coil-controller will separate coil-ipam-controller and coil-egress-controller.
I believe that there are no behavior changes, but some resources(for example, ServiceAccount and ValidatingWebhookConfiguration) are affected by this change.

We also have to consider how to migrate the controller separation.

Now, I'm not sure I work on this as another PR.
Anyway, please wait for a moment to make a decision.

terassyi commented:
@p-strusiewiczsurmacki-mobica 
I'm sorry for late.

I tested the compatibility from 2.8.0 to this branch.

You can see its test in the following repository.
https://github.com/terassyi/terakoya/tree/coil-compatibility-test/kindcluster/coil-stand-alone-egress-compatibility

As a result, we can update this branch without any disruption.
However, to migrate, we have to delete the following resources:

- deployment/coil-controller
- validatingwebhookconfiguration/coilv2-validating-webhook-configuration
- mutatingwebhookconfiguration/coilv2-mutating-webhook-configuration

So, we will provide the migration method, such as documents or some programs. 

Anyway, we can merge this!
Thanks a lot!
